simpleParser.go

```diff
@@ -15,12 +15,19 @@ import (
  * Values can be string or floats. Strings should be enclosed by double quotes, as to not be confused
  * with field variables
  *
+ * Field variables should be encased by backticks `, and embedded objects are accessed via (.) accessor
+ * Arrays are accessed with brackets, and integer indexes do not have to be enclosed by backticks.
+ * Embedded objects can also be accessed via brackets, but must be encased by backticks.
```

Comment:
is there a particular reason for this? it seems to be simpler and more consistent without this option

Reply:
I agree, the use of backticks should probably be optional and only be required in cases where it is needed to disambiguate the grammar (for instance, when a field name itself contains a `.`).

Reply:
actually my comment was not toward the use of backticks. I do agree that backticks can be made optional, only in the case you mentioned.

my comment was on the last line before it: "embedded objects can also be accessed via brackets". I do not see a clear benefit to this. it seems to be simpler and less confusing without this option.

Reply:
Ahh fair enough.  I actually think that allowing access to embedded objects via brackets is important, as it allows us to implement dynamic access to fields in an object in the future (although I'm not even sure if this is possible with the JSONSM algorithm as it stands today, since it center's on the idea of being able to decompose the expressions referenced fields).  It IS worth noting that use of brackets should not use backticks, but rather normal strings (I think).  For instance:

Bad:
```
users[brett19]
users[`brett19`]
"users"["brett19"]
'users'["brett19"]
`users`[`brett19`]
```

Good:
```
users["brett"]
`users`["brett19"]
`users`['brett19']
users[1]
```

Reply:
My understanding is that array index reference and embedded obj references are the same. It all ends up the same in the output path, but I concur that having fewer option may be better in this case.

Reply:
"I actually think that allowing access to embedded objects via brackets is important, as it allows us to implement dynamic access to fields in an object in the future"
I am not sure what this means. 
Say that we want to access the embedded "address" object within "user" object. We could do this using the dot notation:
user.address
or the bracket notation:
user[address]
The input "address" is statically defined in both options.  I cannot see how one option is more "dynamic" than the other. Did I miss anything?

Reply:
Imagine the following expression: ($doc.friends[$doc.bestFriendIdx])

Reply:
i assume that bestFriendIdx is an integer, and the expression means to retrieve the field corresponding to bestFriendIdx?
IMO, the correct way to phrase this is that, the token between "[" and "]" has to be an integer. It could be an integer literal, or an integer field. It cannot be an embedded object.

My interpretation of "an embedded object can be accessed via brackets" is that, if object A has an embedded object B, B can be accessed through A[B].  This is not what your expression does, either, since doc.bestFriendIdx is not an embedded object or field of doc.friends.

Reply:
It could also be a string, considering (as discussed above) indexing can also get a field.  For instance:
JSON:
```json
{
  "friendIds": {
    "Yu": 1,
    "Brett": 14,
    "Neil": 9
  },
  "bestFriendsName": "Brett",
}
```
Expression: `$doc.friendIds[$doc.bestFriendsName]`
Resolves To: 14

Reply:
Note that JSONSM today cannot handle dynamic indexing like this, but I have thoughts on how to make it possible if its needed.

Reply:
ok. this is really saying that fields of an embedded object can be accessed via brackets. 

As for your previous example, $doc.friends[$doc.bestFriendIdx], was "doc.friends"  meant to be an embedded object as well? I thought that it was an array and the expression would retrieve an element from the array.

I propose that we leave brackets strictly for array indexing, and use some other constructs for dynamic field representation. This way there should be less confusions.